src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -110,31 +110,46 @@ public static string GetAsciiOrUTF8StringNonNullCharacters(this ReadOnlySpan<byt
 
         public static string GetRequestHeaderString(this ReadOnlySpan<byte> span, string name, Func<string, Encoding?> encodingSelector)
         {
+            string result;
             if (ReferenceEquals(KestrelServerOptions.DefaultHeaderEncodingSelector, encodingSelector))
             {
-                return span.GetAsciiOrUTF8StringNonNullCharacters(DefaultRequestHeaderEncoding);
+                result = span.GetAsciiOrUTF8StringNonNullCharacters(DefaultRequestHeaderEncoding);
             }
-
-            var encoding = encodingSelector(name);
-
-            if (encoding is null)
+            else
             {
-                return span.GetAsciiOrUTF8StringNonNullCharacters(DefaultRequestHeaderEncoding);
-            }
+                var encoding = encodingSelector(name);
 
-            if (ReferenceEquals(encoding, Encoding.Latin1))
-            {
-                return span.GetLatin1StringNonNullCharacters();
+                if (encoding is null)
+                {
+                    result = span.GetAsciiOrUTF8StringNonNullCharacters(DefaultRequestHeaderEncoding);
+                }
+                else
+                {
+                    if (ReferenceEquals(encoding, Encoding.Latin1))
+                    {
+                        result = span.GetLatin1StringNonNullCharacters();
+                    }
+                    else
+                    {
+                        try
+                        {
+                            result = encoding.GetString(span);
+                        }
+                        catch (DecoderFallbackException ex)
+                        {
+                            throw new InvalidOperationException(ex.Message, ex);
+                        }
+                    }
+                }
             }
 
-            try
+            // Non-tab control characters are considered invalid
+            if (((ReadOnlySpan<char>)result).IndexOfAny('\u000A', '\u000D') != -1)
```

Comment:
```suggestion
            if (((ReadOnlySpan<char>)result).IndexOfAny('\r', '\n') != -1)
```

Reply:
Can we skip this for HTTP/1.1?

Reply:
```suggestion
            if (((ReadOnlySpan<char>)result).IndexOfAny('\u000A', '\u000D') >= 0)
```
Micro-optimization (on top of @Tratcher's suggestion) -- at the cpu-level this can fuse with the `if` to one instruction.

Reply:
> Can we skip this for HTTP/1.1?

Agreed HTTP/1.1 has already enforced this in its parser

Reply:
Updated to skip for HTTP1